rtrlib/bgpsec/bgpsec.c

```diff
@@ -0,0 +1,635 @@
+			      SKI_SIZE +
+			      sizeof(tmp_sig->sig_len) +
+			      SECURE_PATH_SEG_SIZE;
+
+		/*
+		 * From a certain position on, read until the end of the stream.
+		 * The starting position is determined by the offset. The end
+		 * of the stream is the total length minus the offset.
+		 * The curr pointer points to the resulting "sub-stream".
+		 * Hacky, could be improved.
+		 */
+		uint16_t len = (uint16_t)(get_stream_size(s) - offset);
+
+		uint8_t *curr = lrtr_malloc(len);
+
+		read_stream_at(curr, s, offset, len);
+
+		retval = hash_byte_sequence(curr,
+					    len,
+					    data->alg,
+					    &hash_result);
+
+		lrtr_free(curr);
+
+		if (retval != RTR_BGPSEC_SUCCESS)
+			goto err;
+
+		/* Store all router keys for the given SKI in tmp_key. */
+		unsigned int router_keys_len = 0;
+		enum spki_rtvals spki_retval = spki_table_search_by_ski(
+							table,
+							tmp_sig->ski,
+							&tmp_key,
+							&router_keys_len);
+		if (spki_retval == SPKI_ERROR) {
+			retval = RTR_BGPSEC_ERROR;
+			goto err;
+		}
+
+		/* Loop in case there are multiple router keys for one SKI. */
+		for (unsigned int j = 0; j < router_keys_len; j++) {
+			/* Validate the siganture depending on the algorithm
+			 * suite. More if-cases are added with new algorithm
+			 * suites.
+			 */
+			if (data->alg == RTR_BGPSEC_ALGORITHM_SUITE_1) {
+				retval = validate_signature(hash_result,
+							    tmp_sig,
+							    &tmp_key[j]);
+			} else {
+				retval = RTR_BGPSEC_UNSUPPORTED_ALGORITHM_SUITE;
+				goto err;
+			}
+			/* As soon as one of the router keys produces a valid
+			 * result, exit the loop.
+			 */
+			if (retval == RTR_BGPSEC_VALID)
+				break;
+		}
+		lrtr_free(hash_result);
+		lrtr_free(tmp_key);
+		hash_result = NULL;
+		tmp_key = NULL;
+		tmp_sig = tmp_sig->next;
+	}
+
+err:
+	if (hash_result)
+		lrtr_free(hash_result);
+	if (tmp_key)
+		lrtr_free(tmp_key);
+	if (s)
+		free_stream(s);
+
+	if (retval == RTR_BGPSEC_VALID)
+		BGPSEC_DBG1(
+			"Validation result for the whole BGPsec_PATH: valid");
+	else if (retval == RTR_BGPSEC_NOT_VALID)
+		BGPSEC_DBG1(
+			"Validation result for the whole BGPsec_PATH: invalid");
+
+	return retval;
+}
+
+int rtr_bgpsec_generate_signature(
+			  const struct rtr_bgpsec *data,
+			  uint8_t *private_key,
+			  struct rtr_signature_seg **new_signature)
+{
+	/* The signature generation result. */
+	enum rtr_bgpsec_rtvals retval = 0;
+
+	/* The resulting hash. */
+	unsigned char *hash_result = NULL;
+
+	/* The required size for the signature */
+	int req_sig_size = 0;
+
+	/* A stream that holds the data that is hashed */
+	struct stream *s = NULL;
+
+	/* Total size of required space for the stream */
+	unsigned int stream_size = 0;
+
+	/* OpenSSL private key structure. */
+	EC_KEY *priv_key = NULL;
+
+	/* Check, if the parameters are not NULL except for *new_signature,
+	 * which must be NULL.
+	 */
+	if (!data || !data->path || !private_key || *new_signature)
+		return RTR_BGPSEC_ERROR;
```

Comment:
This error code is inconsistent with `rtr_bgpsec_validate_as_path`, please also the comment about the error code there.

Reply:
fixed